DietPi-Imager | Add option to add FAT partition after rootfs to simplify setup

dirkhh opened the pull request:
This isn't complete, yet (MBR only, only tested on one board), but I felt that this would be enough to maybe get some feedback on the code and the overall idea for implementation.

What this does right now is that if the corresponding environment variable `ADD_DOS_PART` is set, at image creation time a 4MB DOS partition named DIETPISETUP (yay for 11 character limit) is created at the end of the disk image and `dietpi.txt` and `dietpi-wifi.txt` are copied onto that partition.

During first boot, when such a partition is found, those two files (and ONLY those two files - not sure if that is the right choice, but it seemed better that way) get copied back over the ones that were in the `root` filesystem.

This way a user on a Mac or a Windows system can very easily make changes to the settings - I think the most important use case will be wifi setup on headless systems... and then of course fully automated setup.

Again, I'm curious to hear comments about everything. Coding style, decisions about how this is enabled, where I inserted the code, anything. I want to make sure that this is broadly useful and not just a fix for my use case.

dirkhh committed:
introduce an option to add a trailing fat partition

So far this only handles dos partition tables.
When ADD_DOS_PART is set to 1, add a small 4MB fat filesystem as the
last partition to this image, and copy the two DietPi config files to
this partition.
Now even people who don't have access to a Linux system can very easily
mount the DIETPISETUP partition, edit the two files and get the setup
they want. This is especially important for headless devices connecte
via WiFi - where this initial setup is necessary for them to connect to
the network in the first place.

This commit only creates that new partition. The setup logic still needs
to be updated so that it mounts that partition (when present) and uses
the config files on that partition instead of the ones provided
elsewhere in the image.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
increase roof size for NanoPiNEO3

Sometimes the buids are running out of space when doing a kernel update.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
copy user config on first boot and delete setup partition

This is the second part of the logic. If we have the DIETPISETUP partition then
copy the dietpi.txt and dietpi-wifi.txt files from there into the /boot directory
and delete the DIETPISETUP partition (in order to allow the resizing of the root
filesystem to work).

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng commented on .build/images/dietpi-imager:
```diff
@@ -493,7 +502,34 @@
 			G_EXEC partx -u "$FP_SOURCE"
 		fi
-
+		if [[ "$ADD_DOS_PART" == "1" &&  "$PART_TABLE_TYPE" == 'dos' ]]
```
It is generally also possible with GPT partitions. All tools used support both. `parted` could be used to be failsafe (as in `dietpi-build`), or `sgdisk` to use a GPT partitioning tool explicitly. But even `sfdisk` should work just fine:
```suggestion
		if [[ $ADD_DOS_PART == 1 ]]
```

dirkhh replied:
I know that it is possible with GPT partition tables. It's just that my code so far has only been tested with MBT and  I didn't want to submit untested code.

For your suggested change... shell quoting rules are... well. I'll take your word for it.

MichaIng replied:
Within bash-syntax double brackets `[[ ... ]]`, left hand variables never need to be quoted. Right side usually yes, but for integers we skip them. Although consequently right side could be quoted as well, but then single quotes for literal strings.

We can quickly test it with GPT, also e.g. with the GitHub workflows. Lets add support for it with the same PR.

dirkhh replied:
TIL (today I learned...)
Will fix the syntax. Will work on the GPT part. I need to check which of the dozen or so boards that I have laying around here from the ADSB Feeder testing uses GPT... Orange Pi 5+ maybe? Yep, good guess. So easy enough to test for me.

What is your deadline for hitting a 'next release'... again... I've been using DietPi for a while, but am new to the development side of it, so I want to make sure I understand your expectations.

nonnorm replied:
A release happens every 4 weeks, so the next one will be on September 23rd. The beta is generally released about a week before that (September 16th). Big changes like this are usually merged before the beta release, but it's flexible, and @MichaIng has the final say.

MichaIng commented on .build/images/dietpi-imager:
```diff
@@ -493,7 +502,34 @@
+			export lp_start=$(echo "$old_table" | jq ".partitiontable.partitions[].start" | sort -n | tail -1)
+			# jq syntax is a bit werid: pick the partition with the given start, and extract its size
+			local lp_size=$(echo "$old_table" | jq '.partitiontable.partitions|map(select(.start==(env.lp_start|tonumber)))[].size')
+			local np_start=$(( $lp_start + $lp_size ))
```
See some lines below where we derive `$IMAGE_SIZE` from the last partition's end, which seems to be a simpler approach than parsing JSON. The first of these two lines without the single added sector (?) could be moved up, so the variable is available for this new block and can be increased by 4 MiB. Then the single sector and for GPT additional 33 sectors can be added below.

dirkhh replied:
Partitions can be out of order. But maybe we know that this isn't the case here since we created the partitions? I figured 'better safe than sorry' - but I'm happy to simplify this.
The single added sector is an old habit - there are some odd off by one errors that you can prevent from doing bad things to your data by having a spare sector between them.
Again, happy to remove that to be in line with your preferences.

MichaIng replied:
So what you do is scraping the partition with the highest start sector, regardless which partition number it has/at which position `sfdisk` lists them? Generally it makes sense, also since `dietpi-imager` is generally aimed to work as standalone script. But we do already assume that the root partition is the last partition, at least we minimise the last partition, and an out of order partition table might cause problems elsewhere already. If we want to deal with this properly, then I'd rather do this for all parts of the script, and also DietPi itself (the partition and filesystem expansion on first boot) instead of for this individual feature.

About the single added sector: I am just currently not sure whether `sfdisk` adds new partitions at including the start sector of from the given start sector on. With `parted` you can use the same sector as end sector for a previous and start sector for the next partition, but not sure whether this is the same for `sfdisk`. Generally I'd avoid a one sector gap. Actually I do not see where you add 1 sector in the new block? We do add it at the end of the whole image file after the last partition only (AFAIK there it was needed in some case), respectively 34 sectors for GPT for the backup partition table.

dirkhh replied:
> So what you do is scraping the partition with the highest start sector, regardless which partition number it has/at which position `sfdisk` lists them? 

Yes - as I mentioned elsewhere - I'm not a super bash programmer... but I love the swiss army knife tool chains out there. Like `sfdisk`/`json`/`jq`

> Generally it makes sense, also since `dietpi-imager` is generally aimed to work as standalone script. But we do already assume that the root partition is the last partition, at least we minimise the last partition, and an out of order partition table might cause problems elsewhere already. If we want to deal with this properly, then I'd rather do this for all parts of the script, and also DietPi itself (the partition and filesystem expansion on first boot) instead of for this individual feature.

I noticed that assumption that root is last... and ironically, I noticed that when I use this `add-fat` feature, I clearly break an assumption in the first boot and even though the fat partition is removed very early on, the root partition still didn't get resized in my testing...

I would love to see a more generic solution to this. If you give me some pointers where this should be implemented and what API you'd be looking for, I'd be happy to tackle that as well.

> About the single added sector: I am just currently not sure whether `sfdisk` adds new partitions at including the start sector of from the given start sector on. With `parted` you can use the same sector as end sector for a previous and start sector for the next partition, but not sure whether this is the same for `sfdisk`. Generally I'd avoid a one sector gap. Actually I do not see where you add 1 sector in the new block? We do add it at the end of the whole image file after the last partition only (AFAIK there it was needed in some case), respectively 34 sectors for GPT for the backup partition table.

I looked again and it turns out that I removed that. I am now using `local np_start=$(( $lp_start + $lp_size ))` - I originally had a `+ 1` in there, but apparently decided to remove that before submitting the PR 🙃
And yes - I see that you use "magic just add the partition" syntax elsewhere. I generally prefer to tell `sfdisk` exactly what I want and where I want it... fewer nasty surprises that way. But again - happy to follow your preferences...

MichaIng commented on .build/images/dietpi-imager:
```diff
@@ -338,6 +338,15 @@
+			cp "$FP_MNT_TMP"/boot/dietpi.txt "$TMP_DIETPI_CONF_DIR"/
+			cp "$FP_MNT_TMP"/boot/dietpi-wifi.txt "$TMP_DIETPI_CONF_DIR"/
+			G_EXEC_OUTPUT=1 G_EXEC ls -l "$TMP_DIETPI_CONF_DIR"/
+		fi
```
I suggest to just mount the rootfs and the new partition both in the same block below. I see no reason to copy the config files separately into a tmp dir here, and then from there to the new fs below. The rootfs could be mounted R/O down there to rule out any accidental writes.

Also the still existing loop device could be used instead of removing it first and creating a new one.
```sh
losetup -c "$FP_SOURCE"
```
can be used to re-read the image size. Also note that at this point it is only needed if `[[ $FP_SOURCE_IMG ]]`, else the source is a drive, `dd` can/must be skipped and the new partition mounted directly after `partprobe`/`partx -u`.

dirkhh replied:
I'll need to look into that comment when I had more coffee (it's only 7am here). I think I know what you are telling me. Simplify things by mounting the rootfs r/o, reusing the existing loop device.
And pay attention to the case where source is a drive.

MichaIng replied:
Right, and having all code related to this feature merged into a single code block 🙂. Enjoy your coffee and breakfast.

dirkhh committed:
rewrite imager changes according to Micha's feedback

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
make the copy / delete / resize at boot work

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
ensure that sfdisk sees the new image size

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
make the fat partition detection work on gpt

We were checking for the mbr type of 'c'. Instead of checking for that
and the GUID for a generic MS data partition, make things easier and
just check that it's a vfat filesystem. With that, the 4MB, and the
subsequent check for the label this appears to be enough belts and
suspenders.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng commented on rootfs/var/lib/dietpi/services/fs_partition_resize.sh:
```diff
@@ -39,6 +39,26 @@
+	then
+		# the last partition is a 4M FAT filesystem - let's check if it is ours
+		SETUP_PART=$(sfdisk -l "$ROOT_DRIVE" | tail -1 | mawk '{print $1}')
+		if blkid "$SETUP_PART" | grep 'LABEL="DIETPISETUP"'
```
Shall we make it a little simpler and just check for the label, using one which is as unique as possible? "DIETPISETUP" is probably sufficiently unique, otherwise "TMPDIETPITXTPART" or something?

Generally this can be used to do all checks in a single line:
```sh
[[ $(lsblk -nrbo FSTYPE,SIZE,LABEL /dev/mmcblk0 | tail -1) == 'vfat 4194304 DIETPISETUP' ]]
```
But maybe we should/can skip the size when using a sufficiently unique label, so we can change it later in a single place when needed. E.g. currently 1 MiB would be sufficient, but we might want to add the possibility to add further configs files or flags there at a later time. I have some in mind we should handle already now, if present: `dietpiEnv.txt`, `config.txt`, `unattended_pivpn.conf`, `Automation_Custom_PreScript.sh`, `Automation_Custom_Script.sh`.

And probably we should add a readme which briefly explains the partition? A little note that one must take care to preserve/use UNIX/LF line endings in text files might be helpful as well, or we convert them on copy?

dirkhh replied:
> Shall we make it a little simpler and just check for the label, using one which is as unique as possible? "DIETPISETUP" is probably sufficiently unique, otherwise "TMPDIETPITXTPART" or something?

You have only 11 characters.
That's why I picked this. It's 11 characters and tells you what it's for.

> Generally this can be used to do all checks in a single line:
> ```sh
> [[ $(lsblk -nrbo FSTYPE,SIZE,LABEL /dev/mmcblk0 | tail -1) == 'vfat 4194304 DIETPISETUP' ]]
> ```

Neat. I like it much better than my attempt.

> But maybe we should/can skip the size when using a sufficiently unique label, so we can change it later in a single place when needed. E.g. currently 1 MiB would be sufficient, but we might want to add the possibility to add further configs files or flags there at a later time. I have some in mind we should handle already now, if present: `dietpiEnv.txt`, `config.txt`, `unattended_pivpn.conf`, `Automation_Custom_PreScript.sh`, `Automation_Custom_Script.sh`.
> 
> And probably we should add a readme which briefly explains the partition? A little note that one must take care to preserve/use UNIX/LF line endings in text files might be helpful as well, or we convert them on copy?

All excellent points. I picked 4MB almost randomly. You are right that 1MB could hold most things that we'd ever consider.
And adding the Readme also makes perfect sense.

Ironically, I don't have easy access to a Windows system, so I don't know if it would mangle like endings. It might be better to "fix it in post", i.e. make sure this is correct when copying.

I'm paranoid, so I'd still want to have a fixed list of files that are supported. Is the list above what you'd suggest?

MichaIng replied:
On common editors on a non-ancient Windows, LF line endings are preserved. We had an automatic fix for `dietpi.txt` as failsafe step in the past for the boot FAT partition (where present) but I removed it a long time ago and remember only a single case in years where one used an unusual editor. It didn't seem worth doing file manipulations. However, here we do file copy/writes anyway, this feature explicitly addresses Windows (and macOS) users, and as of the other optional files one might create new ones, where Windows editors pick CRLF endings by default.

I fully agree with a fixed list of supported files, so jep, the above list should be complete for now, and those are present either depending on the hardware (like `config.txt` only on RPi, ... wait, there we do not need that FAT, remove it from the list 😅) or optional for additional first boot install automation.

MichaIng commented on .build/images/dietpi-imager:
```diff
@@ -493,7 +494,39 @@
+			local root_part=$(sfdisk -qlo DEVICE "$FP_SOURCE" | tail -1)
+			local np_start=$(($(sfdisk -qlo END "$FP_SOURCE" | tail -1) + 1))
+			local type='EBD0A0A2-B9E5-4433-87C0-68B6B72699C7'
+			[[ $PART_TABLE_TYPE == 'dos' ]] && type='c'
```
Did you test whether the `+1` is needed with `sfdisk`?

It should not be needed to add a hardcoded GUID here, is it?
```suggestion
			local np_start=$(($(sfdisk -qlo END "$FP_SOURCE" | tail -1) + 1))
			local type=
			[[ $PART_TABLE_TYPE == 'dos' ]] && type=',type=c'
```
And then `echo "start=$np_start,size=8192$type"` below?

dirkhh replied:
> Did you test whether the `+1` is needed with `sfdisk`?

Yes. If you give it the end sector of the previous position it will error it with "sector is already used"

> It should not be needed to add a hardcoded GUID here, is it?

I believe it is. You need to mark this as a Windows partition so it gets mounted when plugged in on macOS or Windows.
And that crazy GUID is just the market for a Windows data partition.
Of course I could be reading the documentation wrong. 

> ```suggestion
> 			local np_start=$(($(sfdisk -qlo END "$FP_SOURCE" | tail -1) + 1))
> 			local type=
> 			[[ $PART_TABLE_TYPE == 'dos' ]] && type=',type=c'
> ```
> And then `echo "start=$np_start,size=8192$type"` below?

So what kind of position would this get you on gpt? My reading of the docs seemed to imply that by default you'd get a Linux partition. Which Windows or MacOS won't touch  regardless of the filesystem on it. But that's easy to try once I had breakfast 😁

MichaIng replied:
Ah you are right, I mixed up the UUID with the GUID somehow:
- https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_type_GUIDs
- https://en.wikipedia.org/wiki/Microsoft_basic_data_partition

dirkhh committed:
enable DOS setup partition unless it's unnecessary

And collect all the code for this in one spot.
Putting this all in one line is of course possible, but also much harder
to read, understand, and maintain, IMHO.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
switch to 1MB fs size and simplify the code

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
add readme to the extra dos partition

Also copy the dietpiEnv.txt file. The others don't appear to have
default content, so I just mention them in the readme.

Frankly, the text in the readme probably needs some editing.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
remove the root filesystem size changes

These were needed for me to be able to test things locally, but don't
belong in this PR.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
cleanups

comparing to dev this removes an odd whitespace change; it also tightens
some of the code to be more in line with the rest of the file, and
remove some remaining debug output that otherwise just clutters the user
experience when manually building an image.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng committed:
v8.22

- DietPi-Imager | Turn ADD_DOS_PART env variable into a CLI flag. Do not add the additional partition by default, but only if requested
- DietPi-Imager | Coding: Re-use FP_SOURCE_IMG and IMAGE_SIZE variables for adding FAT patition; only raise source image size if required
- DietPi-Imager | Copy as well dietpiEnv.txt and optional config/automation files/scripts to setup FAT partition if present
- DietPi-FS_partition_resize | Allow FAT setup partitions of any size and only copy files if newer: Reason is that users with a Linux host may have edited/created files on the ext4 partition and we do not want to overwrite these
- DietPi-Build | Make skipping the additional FAT setup partition a CLI flag, and pass the new CLI flag to DietPi-Imager
- DietPi-Build | When creating multiple image editions for the same device, always remove the added FAT setup partition before booting the system or passing it to DietPi-Imager again
- CHANGELOG | Add entry about added FAT setup partition
